### ISSUE TITLE
Tests failing on SSHFS

### ISSUE THREAD

**Issue by PlasmaPower** [borgbackup/borg#1362]
All of these issues have been fixed in #1373.

- [ ] No ACLs: `KeyError: 'acl_access'` pops up a bunch - also present on ntfs-3g (see #1363)
- [ ] No FIFOs:
  ```
          # FIFO node
>       os.mkfifo(os.path.join(self.input_path, 'fifo1'))
E       PermissionError: [Errno 1] Operation not permitted

.tox/py35/lib/python3.5/site-packages/borg/testsuite/archiver.py:290: PermissionError
  ```

- [ ] No unix sockets:
   ```
            sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
>       sock.bind(os.path.join(self.input_path, 'unix-socket'))
E       PermissionError: [Errno 1] Operation not permitted

.tox/py35/lib/python3.5/site-packages/borg/testsuite/archiver.py:377: PermissionError
   ```

- [ ] No hardlinks:
   ```
                self.cmd('extract', self.repository_location + '::test', '--strip-components', '2')
>           assert os.stat('hardlink').st_nlink == 2
E           AssertionError: assert 1 == 2

.tox/py35/lib/python3.5/site-packages/borg/testsuite/archiver.py:563: AssertionError
   ```

These should be fixed before #1359 is merged.

**Comment by PlasmaPower** [comment]
To solve these issues, should we check the name of the FS or try to detect which features it supports?

**Comment by enkore** [comment]
Feature detection is preferable imho.

**Comment by PlasmaPower** [comment]
Of course, but it would be substantially harder, especially with #1363 with the permission and mode stuff.

Maybe have flags which we can manually set per FS name?

**Comment by PlasmaPower** [comment]
Maybe we could test for support of ACLs, hard links, unix sockets, and FIFOs, but use the name to determine if permissions and modes are reasonable?

**Comment by enkore** [comment]
Yeah, the errors you got are in create_test_files (or setUp) or so anyway, which also does the have_root detection.

I don't know any simple/specific way (a syscall) to get the FS name, so parsing should be conservative I think.

**Comment by PlasmaPower** [comment]
@enkore I think we could parse `/proc/mounts` which shouldn't be too bad (it's meant to be parsed).

**Comment by enkore** [comment]
Looking at the couple different systems here (solaris, oi, bsds) I'd say neither mount(8) nor /proc/mounts is a feasible idea.

Since this is mostly about the FS tests for vagrant I'd say it would be simpler and far more robust to just explicitly disable tests that would fail there.

**Comment by PlasmaPower** [comment]
So maybe pass in an ENV variable `DISABLE_FS_MODE_TESTS`?

**Comment by enkore** [comment]
No need, you should be able to pass them directly through to py.test [through tox](http://tox.readthedocs.io/en/1.0/example/general.html#interactively-passing-positional-arguments). Something like `tox ... -- borg.testsuite -k 'not ...' -k 'not ...'`

**Comment by PlasmaPower** [comment]
I'm not sure specific test names should be in the test driver script, especially given the amount of tests. We could consider putting them in a separate text file, but I think they'd fit better in the code.

**Comment by enkore** [comment]
These four here can be done with feature detection, the others well... the borg.testsuite doesn't really need to know about more exotic FSes at least not by name. So it would seem reasonable to me to keep the test-skip-list in the script used to test exotic FSes.

**Comment by PlasmaPower** [comment]
@enkore There are a lot of tests that fail with bad permissions or modes though. I'd like an ENV variable `DISABLE_FS_MODE_TESTS`, and then we could do `@unittests.skipif("os.environ['DISABLE_FS_MODE_TESTS']")`. I think it makes more sense to put the disablers next to the tests than in a completely different file, it will help when tests change.

**Comment by enkore** [comment]
Makes sense. Avoids putting all tests on a list somewhere else, but doesn't put very unportable mount-detection code in the testsuite :)

**Comment by PlasmaPower** [comment]
I'm working on properly skipping and iffing, but it might be a bit before I finish it. I'm going to make one PR for all the subissues of this issue, then another PR for `ntfs-3g`'s funky permissions and modes. Those together should fix everything.

**Comment by PlasmaPower** [comment]
99 acl errs in the log, 99 acl errs
Take one down, skip it around, 99 acl errs in the log

On another note, on SSHFS `os.readlink('/path/to/symlink')` works great but `os.stat('/path/to/symlink')` throws a FileNotFoundError. If I wasn't lazy, I'd submit a bug report to SSHFS.

**Comment by PlasmaPower** [comment]
`noatime`'s failing on SSHFS, and that's with the changes in master. Why do we even have this check if has_noatime returns true? Shouldn't we just skip it in that case?

**Comment by enkore** [comment]
[This assert?](https://github.com/borgbackup/borg/blob/master/src/borg/testsuite/archiver.py#L420)

**Comment by PlasmaPower** [comment]
@enkore No, line 417. Basically it's asserting that `has_noatime` is correct, rather than anything useful. If `has_noatime` returns true, we should just `pytest.skip('filesystem does not have atime')`.

**Comment by enkore** [comment]
has_noatime returns true if O_NOATIME is around *and* works (open-read-close cycle doesn't change the atime), or should do that... in L417 the assertion is that the atime of the input file did not change, and that the output file also has the correct atime, ie. that borg correctly used O_NOATIME.

L417 should not be able to fail if has_noatime returned true.

**Comment by PlasmaPower** [comment]
@enkore Oh, I see. That makes sense. Hmm, I think the problem with SSHFS might be an inability to set atime then. I'll see if I can also test for that.

**Comment by PlasmaPower** [comment]
Does anyone have any idea what's happening with this error?

```python
class ArchiverTestCase(ArchiverTestCaseBase):
    def test_basic_functionality(self):
        have_root = self.create_test_files()
        # fork required to test show-rc output
        output = self.cmd('init', '--show-version', '--show-rc', self.repository_location, fork=True)
        self.assert_in('borgbackup version', output)
        self.assert_in('terminating with success status, rc 0', output)
        self.cmd('create', self.repository_location + '::test', 'input')
        output = self.cmd('create', '--stats', self.repository_location + '::test.2', 'input')
        self.assert_in('Archive name: test.2', output)
        self.assert_in('This archive: ', output)
        with changedir('output'):
>           self.cmd('extract', self.repository_location + '::test')

.tox/py35/lib/python3.5/site-packages/borg/testsuite/archiver.py:391:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox/py35/lib/python3.5/site-packages/borg/testsuite/archiver.py:301: in cmd
    self.assert_equal(ret, exit_code)
E   AssertionError: 1 != 0
#------------------------------------------------------------------------------------ Captured stdout call ------------------------------------------------------------------------------------
input/link1: [Errno 2] No such file or directory

#------------------------------------------------------------------------------------ Captured stderr call ------------------------------------------------------------------------------------
#------------------------------------------------------------------------------
```

It happens pretty much every time a test calls extract. I've confirmed that `input/link1` exists as a symlink to `something`. It seems like it's actually a flaw in Borg. This occurs on SSHFS, but not ext4.

**Comment by enkore** [comment]
Hmm, could just mean that the symlink target doesn't exist ("ENOENT A directory component in linkpath does not exist or is a dangling symbolic link, or target or linkpath is an empty string." -- while at the same time "No checking of target is done.". The 3P manpage also mentions that the target shall not be validated™).

I can't reproduce this here, so possibly a bug in SSH or sshfs that was fixed?

```py
>>> import os
>>> os.symlink('target does not exist', 'some_symlink')
>>>
```

```
$ sshfs -V
SSHFS version 2.7
FUSE library version: 2.9.7
fusermount version: 2.9.7
using FUSE kernel interface version 7.19
```

```
OpenSSH_7.2p2, OpenSSL 1.0.2h  3 May 2016
```

**Comment by PlasmaPower** [comment]
Well this issue probably isn't specific to the tests, correct? If that's the case, then we probably shouldn't have a test specific solution if we want Borg to work with sshfs. If your solution 1 doesn't cause a cascading error, I'd prefer that.

Do you know what filesystem operation is causing that error? It's not just creating the symlink, we do that for the input just fine. Maybe Borg is actually creating an invalid symlink, but only sshfs throws an error because of it?

**Comment by enkore** [comment]
Found it. It's os.utime failing on sshfs for symlinks.

**Comment by PlasmaPower** [comment]
Sounds good. I'm on mobile now, but once I make that change the sshfs fix pull request should be complete.

**Comment by PlasmaPower** [comment]
Fixed with merging of #1373.
